cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,455 @@
+
+Inbound sessions are not inherently limited in the same way.
+
+It's probably worth considering whether any limits or at least metrics should be put onto inbound + outbound mTLS authentication requests to evaluate the resource impact of the feature.
+
+### Impact: Connection authentication upon policy transition
+Using an auth table with an expiry time means that, on policy transition, the worst-case scenario is that a removed policy would still be valid for the shortest lifetime of the two certificates involved.
+
+That is, there are four scenarios:
+- Connection starts, auth disabled: This is not impacted
+- Connection starts, auth enabled: The flow in the above diagrams is used; packets will be dropped until the mTLS handshake is complete.
+- Connection running, auth disabled -> enabled: Once the policy is seen by the agent and the datapath programmed, packets will be dropped until the mTLS handshake is complete.
+- Connection running, auth enabled -> disabled: Without any action from the userspace, packets will be permitted until the expiry time is hit; it seems likely that a periodic garbage-collection process will need to run in userspace to remove expired and invalid auth map entries.
+
+Either of these impacts could be mitigated by delaying the policy implementation when handing a policy update until all impacted peers are authenticated and the dataplane CT entries are updated to reflect the change in authentication policy.  Careful attention should be paid to the implementation either way to ensure the minimal disruption to existing connections while providing the maximal assurance that connections impacted by a new auth policy are in fact authenticated. If the peers for existing connections cannot be authenticated within a bounded period, these connections should be highlighted and potentially terminated.
+
+### Impact: Connection disruption during Agent downtime
+The current design relies primarily upon the datapath containing a cache of authenticated connections, and the datapath falls back to requesting authentication from the userspace agent when the cache does not indicate a connection is authenticated. This introduces a hard dependency on the availability of the agent for _new_ identity->identity pairings. Note that because this will be done on an _identity_ basis rather than an _IP Address_ basis, multiple connections between the same Pods will _not_ need to pay the TLS session handshake time cost. (Usually on the order of up to 100ms, but it adds up.)
```

Comment:
The `Note that because ...` all the way through the end of the line seems to be unrelated to connections during agent downtime, since it's talking about TLS handshakes which should be impossible during agent outage.

Reply:
I'm also curious where the 100ms number comes from and whether there are aspects of the design that help Cilium keep within that sort of range? (Also whether scale plays a role in being able to stick to those numbers)

Reply:
The 100ms number was a wild-ass guess. I'm just going to remove this bit, as I think you're right and it doesn't make sense for it to be there.

Reply:
For what it's worth, I like the _idea_ behind providing concrete numbers like that, but I think it's just hard to reason about the viability without considerable implementation details being specified here. Maybe something good to follow up on - we should have metrics for this and be able to measure+optimize for it.